### PR TITLE
[WIP] fluentd exporter

### PR DESCRIPTION
**Pull request by maxgolov**
@lalitb - Draft implementation of `fluentd` exporter.

What I'm still working on:

- ability to rename "standard" fields, since semantics of the fields are different in Azure MDSD:
https://github.com/Azure/azure-linux-extensions/commit/5ab581bc336b701371a783a143768951fe8ed630
https://github.com/Azure/WALinuxAgent/issues/148
And I'd like to keep the code entirely vendor-agnostic, i.e. with ability to customize the "standard" field names. This will be done similar to how it's done in ETW exporter, by allowing custom header that provides a set of compile-time constants (macros) that define the field names.

- need to rework the individual event storage model from `json object` to `std::list` or to `std::map<std::string,AttributeValue>`, that'll be better for performance than storing all events in intermediate JSON.

PR includes the test setup needed to run a test fluent agent: `fluentd` or `fluentbit`, both work. It also works with specific of MDSD custom protocol. MDSD itself is available in open Azure Container Registry. I can elaborate on that setup (how to use this exporter to send telemetry to Azure mdsd via fluent) in a separate document.

### CODE REVIEW

**Comment by lalitb** [comment]
Thanks @maxgolov  will have a look.

**Comment by maxgolov** [comment]
I added async uploader with batching queue, configurable reconnection retries, and configurable wait time between batch uploads. Plus E2E tests. E2E test starts up "fluentd-alike" listener, that accepts MsgPack encoded fluentd forward protocol. And verifies that expected number of events have arrived. It also prints the JSON representation of MsgPack in console.

E2E tests depend on another PR in the main repo, that implements tests socket server for TCP, UDP and Unix Domain sockets. Test server included here is not intended to be used in production, for testing only, and fundamentally it's an improvement on top of HTTP server we use elsewhere. Now it supports not only HTTP, but raw TCP, UDP and Unix Domain cross-plat.

File: exporters/fluentd/src/fluentd_exporter.cc [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+          seq_evt_++;
+          LOG_TRACE("exporting: batch #%d evt #%d", seq_batch_, seq_evt_);
+        }
+      }
+    }
+    obj.push_back(spanevents);
+    LOG_TRACE("sending %zu Span event(s)", obj[1].size());
+    std::vector<uint8_t> msg = nlohmann::json::to_msgpack(obj);
+    // Schedule upload of Span event(s)
+    Enqueue(msg);
```
Comment (lalitb) [inline exporters/fluentd/src/fluentd_exporter.cc:155]: So, we are batching the events here, and sending them through a separate thread, to optimize the number of TCP connections? This is not as per the specs, as the `Exporter::Export()` method should really be doing the export through the transmission channels, and the return status should be the actual status of transmission : https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#exportbatch

There is already batch span processor to handle any such batching needs, but shouldn't be here as part of exporter.

File: exporters/fluentd/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   $ cmake -DWITH_fluentd=ON ..
+   -- The C compiler identification is GNU 9.3.0
+   -- The CXX compiler identification is GNU 9.3.0
+   ...
+   -- Configuring done
+   -- Generating done
+   -- Build files have been written to: /home/<user>/source/opentelemetry-cpp/build
+   $
+```
+
```
[the remark this answers is not in the document]
Comment (maxgolov) [reply exporters/fluentd/README.md:24]: No, that is not how it is supposed to be built. It is expected to be built as an overlay on top of main repo.

**Comment by lalitb** [comment]
Closing this PR. Plan is to consolidate the changes in this PR along with the socket library introduced in https://github.com/open-telemetry/opentelemetry-cpp/pull/871 and raise common PR.

**Pull request closed without merging**